timemachine/cpp/src/bound_potential.cu

```diff
@@ -3,12 +3,9 @@
 
 namespace timemachine {
 
-BoundPotential::BoundPotential(std::shared_ptr<Potential> potential, const int size, const double *h_p)
-    : size(size), d_p(nullptr), potential(potential), max_size_(size) {
-    if (this->size > 0) {
-        d_p.reset(new DeviceBuffer<double>(this->size));
-        d_p->copy_from(h_p);
-    }
+BoundPotential::BoundPotential(std::shared_ptr<Potential> potential, const std::vector<double> &params)
+    : size(params.size()), buffer_size_(size), d_p(buffer_size_), potential(potential) {
```

Comment:
Isn't it invalid to construct a DeviceBuffer with size 0? Isn't that why previously this was a `unique_pointer<DeviceBuffer>`?

Reply:
Hmm, what is the use case for passing empty parameters here? IIRC the restriction against initializing `DeviceBuffer(0)` is artificial and it could just be a no-op?

Reply:
Removed the `size > 0` assertion from `DeviceBuffer` in https://github.com/proteneer/timemachine/pull/1119/commits/1c276be74ba253c8a11e75cc1bf245d3b3d305e8. Admittedly I'm not totally sure why we have it, but removing this simplifies `BoundPotential` and possibly other code as well where we allow passing empty arrays. (This also feels more consistent with other array/buffer APIs)

Reply:
> Hmm, what is the use case for passing empty parameters here? IIRC the restriction against initializing `DeviceBuffer(0)` is ad-hoc and it could just be a no-op?

I don't have a clear use case, but the old code supported it and didn't want to remove support for zero length params unless it is clear that we don't require it. @proteneer is there a potential we might want that has no parameters?

Reply:
It's certainly plausible to have potentials that are devoid of parameters, but we wouldn't need to call `set_params` on them anyways.

Reply:
Could we add a test for constructing a BoundPotential with zero length params? Just to make sure this edge case is covered by our tests

Reply:
The `CentroidRestraint` is actually a potential we have that doesn't use any parameters

Reply:
Added a test verifying that the `BoundPotential::BoundPotential` and `BoundPotential::set_params` work as expected when the parameters are an empty array.